cluster-autoscaler/cloudprovider/clusterapi/clusterapi_provider.go

```diff
@@ -86,8 +87,9 @@ func (p *provider) NodeGroupForNode(node *corev1.Node) (cloudprovider.NodeGroup,
 // HasInstance returns whether a given node has a corresponding instance in this cloud provider
 func (p *provider) HasInstance(node *corev1.Node) (bool, error) {
 	machineID := node.Annotations[machineAnnotationKey]
+	ns := node.Annotations[clusterNamespaceAnnotationKey]
 
-	machine, err := p.controller.findMachine(machineID)
+	machine, err := p.controller.findMachine(path.Join(ns, machineID))
```

Comment:
I'm a bit confused how we could update the machine lookup key in this way and not have that break things.

Reply:
i think this would work, unless `ns` was empty perhaps? but if it ever ran on windows it would break. probably better if we just use a `fmt.Sprintf` instead.

Reply:
> I'm a bit confused how we could update the machine lookup key in this way and not have that break things.

To make sure I understood you correctly: are you worried about future refactorings where we'd change how the machine lookup key is created in the informer? 

The question triggered a few questions I had about the informer and how it's set up. 
Looking at the code and the debug output I see the following:

- machineInformer `keyFunc` is `k8s.io/client-go/tools/cache.DeletionHandlingMetaNamespaceKeyFunc`
- it additionally adds an indexer by providerID using `indexMachineByProviderID`

The `keyFunc` uses [ObjectNames](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/object-names.go) to construct the namespace/name string. 

The lookup therefore can be done either by the `providerID` (which we don't have in the corev1.Node object) or by `namespace/name` string. 

Using `path.Join` or `fmt.Sprintf` doesn't matter much, probably. If the `ns` is not set, the machine won't be found. In the clusterapi codebase I saw both `fmt.Sprintf` and `path.Join` being used, for slightly different or similar purposes. 

I'm happy to change or keep it as-is - please let me know your preference.

Reply:
This is gonna sound dumb but I think I didn't realize that the intention of this change was literally to fix a broken thing. Naively I assumed the current lookup worked, and so I didn't grok how it would continue to work w/ the introduction of the namespace prefix. Now I understand (if I understand correctly) that the current lookup implementation is broken!